File: Javascript/SortingAlgorithms/BubbleSort.js
// O(n^2) time, O(1) Space,  Stable Sort: maintain the relative order of records with equal keys (e.g: value)
const BubbleSort = (array) => {
  // Flag that holds info about whether the swap has occur or not.
  let flag = false;
  // Clone original array to prevent its modification.
  const otherArray = [...array];

  for(let i = 0; i < otherArray.length; i++) {
    flag = false;

    for(let j = 0; j < otherArray.length - i; j++) {
      if(otherArray[j] > otherArray[j + 1]) {
        [otherArray[j], otherArray[j + 1]] = [otherArray[j + 1], otherArray[j]];
        
        flag = true;
      }
    }
  }

  // If there were no swaps then array is already sorted and there is
  // no need to proceed.
  if(!flag) {
    return otherArray;
  }
  return otherArray;
}

File: Java/SortingAlgorithms/BubbleSort.java
/**
 * O(n^2) worst, average, best time complexity, O(1) space
 */
public class BubbleSort {
    public static void main(String[] args) {
        int[] nums = new int[]{5,4,3,2,1,6,7,4,5,2,1,3,0};
        BubbleSort bs = new BubbleSort();
        bs.bubbleSort(nums);
        for(int num: nums) {
            System.out.print(num + "   ");
        }
    }

    public void bubbleSort(int[] nums) {
        if(nums == null || nums.length == 0) {
            return;
        }

        for(int i = 0; i < nums.length; i++) {
            for(int j = 0; j < nums.length - i - 1; j++) {
                if(nums[j] > nums[j + 1]) {
                    swap(nums, j, j + 1);
                }
            }
        }
    }

    private void swap(int[] nums, int left, int right) {
        int tmp = nums[left];
        nums[left] = nums[right];
        nums[right] = tmp;
    }
}
